cou/steps/hypervisor.py

```diff
@@ -0,0 +1,164 @@
+# Copyright 2024 Canonical Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Hypervisor planner."""
+import logging
+from dataclasses import dataclass
+
+from cou.apps.base import ApplicationUnit, OpenStackApplication
+from cou.steps import (
+    HypervisorUpgradePlan,
+    PostUpgradeStep,
+    PreUpgradeStep,
+    UpgradePlan,
+)
+from cou.utils.openstack import OpenStackRelease
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class HypervisorMachine:
```

Comment:
Is it too soon to inherit from `Machine`? I guess in the end of the day we will end up with the normal machine class. This hypervisor machine is missing IMO a easy way to access `az`. If we are grouping by AZ, the way it is right now we would need to access units -> unit -> machine -> az

Reply:
I share a similar concern (which is partly what I wanted to express in my initial review comments). IIUC the logic chain is even longer when trying to instantiate a `HypervisorMachines` class: we need to go through `HypervisorUpgradePlanner` -> OpenStackApplications -> units -> unit -> machine -> AZ.

Reply:
Yes, it's since the Machine will change. We can discuss this later when it will be used, these are just pseudo code.

Reply:
Ok, closing to see the future changes.